Lib9c/Action/MimisbrunnrBattle.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                throw new SheetRowNotFoundException(nameof(StageSheet), stageId);
+            }
+            
+            var worldInformation = avatarState.worldInformation;
+            if (!worldInformation.TryGetWorld(worldId, out var world))
+            {
+                // NOTE: Add new World from WorldSheet
+                worldInformation.AddAndUnlockNewWorld(worldRow, ctx.BlockIndex, worldSheet);
+            }
+            
+            if (!world.IsUnlocked)
+            {
+                throw new InvalidWorldException($"{worldId} is locked.");
+            }
+            
+            if (world.StageBegin != worldRow.StageBegin ||
+                world.StageEnd != worldRow.StageEnd)
+            {
+                worldInformation.UpdateWorld(worldRow);
+            }
+            
+            if (world.IsStageCleared && stageId > world.StageClearedId + 1 ||
+                !world.IsStageCleared && stageId != world.StageBegin)
+            {
+                throw new InvalidStageException(
+                    $"Aborted as the stage ({worldId}/{stageId}) is not cleared; " +
+                    $"cleared stage: {world.StageClearedId}"
+                );
+            }
+            
+            sw.Restart();
+            var mimisbrunnrSheet = states.GetSheet<MimisbrunnrSheet>();
+            if (!mimisbrunnrSheet.TryGetValue(stageId, out var mimisbrunnrSheetRow))
+            {
+                throw new SheetRowNotFoundException("MimisbrunnrSheet", stageId);
+            }
+            
+            foreach (var equipmentId in equipments)
+            {
+                if (avatarState.inventory.TryGetNonFungibleItem(equipmentId, out ItemUsable itemUsable))
+                {
+                    var elementalType = ((Equipment) itemUsable).ElementalType;
+                    if (!mimisbrunnrSheetRow.ElementalTypes.Exists(x => x == elementalType))
+                    {
+                        throw new InvalidElementalException(
+                            $"ElementalType of {equipmentId} does not match.");
+                    }
+                }
+            }
+            sw.Stop();
+            Log.Debug("Mimisbrunnr Check Equipments ElementalType: {Elapsed}", sw.Elapsed);
+
+            avatarState.ValidateEquipments(equipments, context.BlockIndex);
+            avatarState.ValidateConsumable(foods, context.BlockIndex);
+            avatarState.ValidateCostume(costumes);
+            
+            sw.Restart();
+            if (avatarState.actionPoint < stageRow.CostAP)
+            {
+                throw new NotEnoughActionPointException(
+                    $"Aborted due to insufficient action point: " +
+                    $"{avatarState.actionPoint} < {stageRow.CostAP}"
+                );
+            }
+            avatarState.actionPoint -= stageRow.CostAP;
+            avatarState.EquipCostumes(costumes);
+            avatarState.EquipEquipments(equipments);
```

Comment:
```suggestion
            avatarState.EquipItems(new List<Guid>(costumes).AddRange(equipments));
```

https://github.com/planetarium/lib9c/blob/6eb98d42ccd4cc7bff86688af6bbb74e01c84d19/Lib9c/Model/State/AvatarState.cs#L522